src/main/kotlin/net/ccbluex/liquidbounce/utils/block/placer/BlockPlacer.kt

```diff
@@ -0,0 +1,433 @@
+        }
+
+        return hasPlaced
+    }
+
+    fun doPlacement(isSupport: Boolean, pos: BlockPos, placementTarget: BlockPlacementTarget) {
+        blocks.removeBoolean(pos)
+
+        // choose block to place
+        val slot = if (isSupport) {
+            support.filter.getSlot(support.blocks)
+        } else {
+            slotFinder()
+        } ?: return
+
+        val verificationRotation = rotationMode.activeChoice.getVerificationRotation(placementTarget.rotation)
+
+        // check if we can still reach the target
+        if (!canReach(placementTarget.interactedBlockPos, verificationRotation)) {
+            return
+        }
+
+        // get the block hit result needed for the placement
+        val blockHitResult = raytraceTarget(
+            placementTarget.interactedBlockPos,
+            verificationRotation,
+            placementTarget.direction
+        ) ?: return
+
+        SilentHotbar.selectSlotSilently(this, slot.hotbarSlot, slotResetDelay)
+
+        if (slot.itemStack.item !is BlockItem || pos.getState()!!.isReplaceable) {
+            // place the block
+            doPlacement(blockHitResult, placementSwingMode = swingMode)
+            placedRenderer.addBlock(pos)
+        }
+
+        targetRenderer.removeBlock(pos)
+    }
+
+    private fun raytraceTarget(pos: BlockPos, providedRotation: Rotation, direction: Direction): BlockHitResult? {
+        val blockHitResult = raytraceBlock(
+            range = wallRange.toDouble(),
+            rotation = providedRotation,
+            pos = pos,
+            state = pos.getState()!!
+        )
+
+        if (blockHitResult != null && blockHitResult.type == HitResult.Type.BLOCK && blockHitResult.blockPos == pos) {
+            return blockHitResult.withSide(direction)
+        }
+
+        if (constructFailResult) {
+            return BlockHitResult(pos.toCenterPos(), direction, pos, false)
+        }
+
+        return null
+    }
+
+    fun canReach(pos: BlockPos, rotation: Rotation): Boolean {
+        // not the exact distance but good enough
+        val distance = pos.getCenterDistanceSquaredEyes()
+        val wallRangeSq = wallRange.toDouble().sq()
+
+        // if the wall range already covers it, the actual range doesn't matter
+        if (distance <= wallRangeSq) {
+            return true
+        }
+
+        val raycast = raycast(range = range.toDouble(), rotation = rotation)
+        return raycast != null && raycast.type == HitResult.Type.BLOCK && raycast.blockPos == pos
+    }
+
+    /**
+     * Removes all positions that are not in [positions] and adds all that are not in the queue.
+     */
+    fun update(positions: Set<BlockPos>) {
+        val iterator = blocks.keys.iterator()
+        while (iterator.hasNext()) {
+            val position = iterator.next()
+            if (position !in positions) {
+                targetRenderer.removeBlock(position)
+                iterator.remove()
+            }
+            blocks[position] = false
```

Comment:
Don't use `operator fun set`, it's for `Map<K, V>`. Use `put` instead.

Reply:
I'll check about that.
Regarding the set, it redirects to `put`.

Reply:
There are two or more different `put` functions for these Maps. `put(Object, Boolean)` and `put(Object, boolean)`. Use the second one to avoid auto packaging.

Reply:
Done (: